### PR TITLE
Slack

### PR DESCRIPTION
**Pull request by pjazdzewski1990**
Will be merged by hand.
Only the first file is for review. The rest is just noise.
I want to push it next Thursday 16.07

### CODE REVIEW

File: _posts/2015-07-07-slack.md [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+- Slack
+---
+
+<img src="https://status.slack.com/img/allgood@2x.png" alt="Slack" style="width: 50%; height:50%"/>
+
+## <a name="motivation" href="#motivation">Motivation</a>
+
+In Scalac we believe that communication is essential in IT and that projects with inefficient information sharing are doomed to fail. Fortunately for us it doesn't always mean that putting people in one place is the only way. More often it's about creating the right spirit for exchanging knowledge and providing ways to do so. As in Scalac we are very often working remotely, we had to come up with a tool to help us communicate. We tried few different products for communication, but in the end we switched to Slack. We are using it for around a year now and are pleased with the rich experience it offers. 
+
+But our ambitions don't stop here. As we are a medium size company we strife for maximum efficiency. We are eager to automate as much of the process as possible, so our employees can spend their time focusing on tasks requiring creativity and accountability. As we are hiring the best people, we don't want them to do boring and repeatable busywork.
```
Comment (mareksuscak) [inline _posts/2015-07-07-slack.md]: Was it really meant "strife" or "strive"? Both have a clever meaning but I'd prefer the latter.

File: _posts/2015-07-07-slack.md [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+But our ambitions don't stop here. As we are a medium size company we strife for maximum efficiency. We are eager to automate as much of the process as possible, so our employees can spend their time focusing on tasks requiring creativity and accountability. As we are hiring the best people, we don't want them to do boring and repeatable busywork. 
+
+It's not always an easy process. You have to face technical difficulties and convince people to trust the new tools. 
+
+How is Slack helping us automate? It's done via Integrations. Slack Integrations contain apps for many useful services as Hangout, Github, Crashlytics or Google Docs. Unfortunately for us it's not enough ... we would like to have a central hub with all our tools, so they have a common interface and are easily accessible. This is where Real Time Messaging API (provided by Slack) and Scala Slack Bot (hacked together by us) come in.
```
Comment (mareksuscak) [inline _posts/2015-07-07-slack.md]: It's called Google Hangouts.

File: _posts/2015-07-07-slack.md [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+First, let's start with configuring the integration in Slack. First go to `https://XXX.slack.com/services/new` (where XXX is the name of your organization). There you can browse all the existing integrations. Search for "Bots" option and click "View" button. Give the new bot a name  and click "Add Bot Integration". After that you will be taken to a page with a detailed view of your new bot. The most important thing is to generate and save the API Token. We will need it for later!
+
+You should see something like this
+
+<img src="https://raw.githubusercontent.com/pjazdzewski1990/scala-slack-bot-example/master/bot-creation.png" alt="New-Bot" style="width: 95%; height:95%"/>
+
+## <a name="first_steps" href="#first_steps">First Steps</a>
+
+Now lets start with the Scala side of things. Create a new Scala application using sbt. Having this in place open your `build.sbt` and add
```
Comment (mareksuscak) [inline _posts/2015-07-07-slack.md]: If lets in this case was meant as "let us" (which probably was) then it should be let's.

File: _posts/2015-07-07-slack.md [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+  override def shutdown(): Unit = {
+    slackBot ! WebSocket.Release
+    system.shutdown()
+    system.awaitTermination()
+  }
+}
+{% endhighlight %}
+[\[BotRunner.scala\]](https://github.com/pjazdzewski1990/scala-slack-bot-example/blob/master/src/main/scala/io/scalac/slack/example/BotRunner.scala)
+
+As Scala Slack Bot is written on top of Akka, we need to create an `ActorSystem` in which the bots will run first. Also we shouldn't forget to close them properly using `sys.addShutdownHook(shutdown())`. Now we can start tinkering with our bots. The heart of the bot system is the ScalaBotActor class that takes few parameters: `BotModules` instance, an `MessageEvenBus` for communication, reference to master class and an `ActorRef` pointing to `UserStorage` actor. You don't have to worry about all of that now. Just concentrate on BotBundle.
```
Comment (mareksuscak) [inline _posts/2015-07-07-slack.md]: 2 mistakes in the following part: "an MessageEvenBus" => use a instead of an and a missing letter "t" so a MessageEventBus works probably better.

File: _posts/2015-07-07-slack.md [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+A note on architecture. As Slack limit the amount integrations you can use. We decided on the following approach. We integrate a single application with Slack, so only one slot is used and you are free to add third-party integrations. On the other hand one application can hold multiple bot modules each providing a distinct feature. Those modules are defined in instances of `BotModules` trait.
```
Comment (mareksuscak) [inline _posts/2015-07-07-slack.md]: As Slack limit the amount integrations you can use. We decided... 

should be

As Slack limits the amount of integrations you can use, we decided...

but overall i would reword also the third sentence in this paragraph because it's not very clear - i.e. if I don't know how many integrations are allowed how can I know what's the advantage of "only one slot is used and you are free to add third-party" - but maybe it's just me

File: _posts/2015-07-07-slack.md [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    override def registerModules(context: ActorContext, websocketClient: ActorRef) = {
+      context.actorOf(Props(classOf[CommandsRecognizerBot], eventBus), "commandProcessor")
+      context.actorOf(Props(classOf[HelpBot], eventBus), "helpBot")
+      context.actorOf(Props(classOf[CalculatorBot], eventBus), "CalculatorBot")
+    }
+  }
+{% endhighlight %}
+[\[BotRunner.scala\]](https://github.com/pjazdzewski1990/scala-slack-bot-example/blob/master/src/main/scala/io/scalac/slack/example/BotRunner.scala#L43)
+
+The body of an `BotModules` is very simple. It defines one function `registerModules` which is responsible for initializing actor instances. One thing to notice here is that the actors share the same instance of `MessageEvenBus` (passed as parameter for easier testing) and all have readable actor names.
```
Comment (mareksuscak) [inline _posts/2015-07-07-slack.md]: an BotModules -> a BotModules

File: _posts/2015-07-07-slack.md [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+      context.actorOf(Props(classOf[HelpBot], eventBus), "helpBot")
+      context.actorOf(Props(classOf[CalculatorBot], eventBus), "CalculatorBot")
+    }
+  }
+{% endhighlight %}
+[\[BotRunner.scala\]](https://github.com/pjazdzewski1990/scala-slack-bot-example/blob/master/src/main/scala/io/scalac/slack/example/BotRunner.scala#L43)
+
+The body of an `BotModules` is very simple. It defines one function `registerModules` which is responsible for initializing actor instances. One thing to notice here is that the actors share the same instance of `MessageEvenBus` (passed as parameter for easier testing) and all have readable actor names. 
+
+In our example we start three modules. The `CommandsRecognizerBot` is a handy utility actor, invisible from the outside, that will translate raw Slack messages for us into a more useful `Command` class. `HelpBot` is the Slacks counterpart of Linux `man` command - it presents users with options and descriptions for commands. The `CalculatorBot` is a dummy bot, that we will write on our own.
```
Comment (mareksuscak) [inline _posts/2015-07-07-slack.md]: is the Slacks -> is the Slack's

File: _posts/2015-07-07-slack.md [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    "/" -> ((x: Double, y: Double) => x/y)
+  )
+
+  override def act: Receive = { }
+}
+{% endhighlight %}
+[\[CalculatorBot.scala\]](https://github.com/pjazdzewski1990/scala-slack-bot-example/blob/master/src/main/scala/io/scalac/slack/example/CalculatorBot.scala)
+
+Let's elaborate on this a little. The `AbstractBot` class is a thin wrapper around Akka's `Actor` class providing few useful utility methods for bot authors and constraints a Bot should meet. There are three things a bot should do: 
+* he should use a `MessageEventBus` for communication
```
Comment (mareksuscak) [inline _posts/2015-07-07-slack.md]: Was supposed to be a list but does not appear to be one after compilation.

File: _posts/2015-07-07-slack.md [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+* he should define a `help` method for maintainers and user
+* and he should provide an `act` method, which behaves roughly the same way as `receive` in an ordinary actor
+
+The first thing you might notice in this draft is `help` being a String => OutboundMessage function. `OutBoundMassage` represents a message that will be sent back to Slack. It takes two parameters: on what channel to send and what to send. Having this basic understanding we can move to implementing the `act` method. We can do this as follows.
```
Comment (mareksuscak) [inline _posts/2015-07-07-slack.md]: Forgotten to use code symbols for "OutboundMessage"?

File: _posts/2015-07-07-slack.md [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+
+      publish(response)
+
+    case Command("calc", _, message) =>
+      publish(OutboundMessage(message.channel, s"No arguments specified!"))
+  }
+{% endhighlight %}
+[\[CalculatorBot.scala\]](https://github.com/pjazdzewski1990/scala-slack-bot-example/blob/master/src/main/scala/io/scalac/slack/example/CalculatorBot.scala#L21)
+
+As you can see this is not far from Akka. The only difference is that we match on `Command` instances provided to us by `CommandsRecognizerBot`. This class carries three values the first is the command name without the leading 'command' character ($ be default), and List[String] of the arguments (without the leading command) and the raw message (so we can, for instance, extract the channel id). After matching a request we perform some computations, form an `OutboundMessage` and pass it to `publish` for sending. Simple as that.
```
Comment (mareksuscak) [inline _posts/2015-07-07-slack.md]: This class carries three values. The first...

File: _posts/2015-07-07-slack.md [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+  }
+{% endhighlight %}
+[\[CalculatorBot.scala\]](https://github.com/pjazdzewski1990/scala-slack-bot-example/blob/master/src/main/scala/io/scalac/slack/example/CalculatorBot.scala#L21)
+
+As you can see this is not far from Akka. The only difference is that we match on `Command` instances provided to us by `CommandsRecognizerBot`. This class carries three values the first is the command name without the leading 'command' character ($ be default), and List[String] of the arguments (without the leading command) and the raw message (so we can, for instance, extract the channel id). After matching a request we perform some computations, form an `OutboundMessage` and pass it to `publish` for sending. Simple as that. 
+
+Few things to consider when writing a new bot. We find it useful to communicate with end users as much as possible about the failures and when a command can't be matched (for instance it's lacking params). You don't have to extend `AbstractBot`, but it's recommended for most cases. If you are in need for creating an internal bot able to communicate with Slack consider using `IncomingMessageListener` directly, but proceed with caution.
```
Comment (mareksuscak) [inline _posts/2015-07-07-slack.md]: I would use "There are few things to consider..." because in it's current form it feels a bit incomplete for me

for most cases -> in most cases

File: _posts/2015-07-07-slack.md [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+For more examples you can check the bots we are using. The source is available [here](https://github.com/ScalaConsultants/scala-slack-bot/).
+
+## <a name="using" href="#using">Using the bot</a>
+
+{% highlight bash %}
+  sbt clean run
+{% endhighlight %}
+
+The command above should start the bot. After all the components start the bot will get online
```
Comment (mareksuscak) [inline _posts/2015-07-07-slack.md]: either use "...components start successfully..." or put a comma after start

File: _posts/2015-07-07-slack.md [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+
+{% highlight scala %}
+api {
+  key = "PUT THE KEY HERE!"
+
+  uuid = "71582001-a791-4ab3-b952-f72d414aab6c"
+  base.url = "https://slack.com/api/"
+}
+websocket.key = "VzvnBtPslX2btvwN3OSp6CuVyeU="
+{% endhighlight %}
```
Comment (marioosh) [inline _posts/2015-07-07-slack.md]: you should also explain what `uuid` and `websocket.key` mean.

File: _posts/2015-07-07-slack.md [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+
+{% highlight bash %}
+  @calculator: calc + 1 2 3
+{% endhighlight %}
+
+## <a name="summary" href="#summary">Summary</a>
+
+<img src="http://www.secretsofthefed.com/wp-content/uploads/2013/07/all_hail_our_new_robot_overlords_by_themarex-d4u5i68.jpg" alt="Bender" style="width: 95%; height:95%"/>
+
+In scalac we have a strong belief than in the future, all redundant tasks will be automated and the done by efficient, specialized robots working on behalf of humans. So people will be able to focus on creative tasks and those tasks that require taking responsibility. We are far from there, yet I'm in the opinion that Slack Bot presented to you above is a step into the future.
```
Comment (mareksuscak) [inline _posts/2015-07-07-slack.md]: 3 mistakes in the first sentence: 
"In scalac we have a strong belief than in the future, all..." -> "In Scalac we have a strong belief that in the future all..."

File: _posts/2015-07-07-slack.md (hunk 9 above)
Comment (marioosh) [inline _posts/2015-07-07-slack.md]: typo in `OutBoundMassage`

**Comment by marioosh** [comment]
After language review LGTM. But... :) Yuo are using "Scalac" or "scalac" in text and Our repository name. Change all of this into **"ScalaC"**

File: _posts/2015-07-07-slack.md (hunk 13 above)
Comment (marekklis) [inline _posts/2015-07-07-slack.md]: I think the same, I don't know what `uuid` & `websocket.key` is and are this parameters necessary.

**Comment by marekklis** [comment]
beside one comment looks good to me :+1:

File: _posts/2015-07-07-slack.md (hunk 13 above)
Comment (pjazdzewski1990) [inline _posts/2015-07-07-slack.md]: @marioosh And what do they mean? ;) I've got only a vague idea about those two.

Comment (marioosh) [inline _posts/2015-07-07-slack.md]: In fact `uuid` and `websocket.key` was the same. But websocket needs in Base64 version. in flow of changes `uuid` wasn't deleted from config. It's no needed for now.

Websocket.key is need to identify your connection. It's Base64 version of GUID placed into header as `Sec-Websocket-Key` for websockets handshaking. (https://tools.ietf.org/html/rfc6455#section-1.3)

File: _posts/2015-07-07-slack.md [hunk 15]
```diff
@@ -0,0 +1,216 @@
+---
+comments: true
+date: 2015-07-16 10:00:00
+layout: post
+slug: scala-slack-bot
+title: Introducing Scala Slack Bot
+summary: In Scalac we use Slack for communication and more. We decided to create a framework for our custom Slack integrations, so we transform it into a living heart of our company. Here are some details on how and why we did that
```
Comment (leszekgruchala) [inline _posts/2015-07-07-slack.md]: Add . at the end :)

File: _posts/2015-07-07-slack.md [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+
+How is Slack helping us automate? It's done via Integrations. Slack Integrations contain apps for many useful services as Google Hangouts, Github, Crashlytics or Google Docs. Unfortunately for us it's not enough ... we would like to have a central hub with all our tools, so they have a common interface and are easily accessible. This is where Real Time Messaging API (provided by Slack) and Scala Slack Bot (hacked together by us) come in.  
+
+The following sections are a step by step guide on how to setup a custom integration for Slack using Scala Slack Bot.
+
+## <a name="configure" href="#configure">Configure Slack</a>
+
+First, let's start with configuring the integration in Slack. First go to `https://XXX.slack.com/services/new` (where XXX is the name of your organization). There you can browse all the existing integrations. Search for "Bots" option and click "View" button. Give the new bot a name  and click "Add Bot Integration". After that you will be taken to a page with a detailed view of your new bot. The most important thing is to generate and save the API Token. We will need it for later!
```
Comment (leszekgruchala) [inline _posts/2015-07-07-slack.md]: `name  and click` - double space

**Comment by leszekgruchala** [comment]
:+1:
